File: Streamoji/Classes/Extensions/UITextView+Emojis.m
//
//  UITextView+Emojis.m
//  Streamoji
//
//  Created by User on 2023/6/7.
//

#import "UITextView+Emojis.h"
#import "EmojiSource.h"
#import "EmojiView.h"
#import "NSAttributedString+Emoji.h"

@interface UITextView ()

@property (nonatomic, strong) UIView *textContainerView;

@end

@implementation UITextView(Emojis)

static NSMutableDictionary<EmojiSource *, UIImageView *> *renderViews;

+ (void)initialize {
    if (self == [UITextView class]) {
        renderViews = [NSMutableDictionary dictionary];
    }
}

- (void)configureEmojis:(NSDictionary<NSString *, id> *)emojis {
    [self configureEmojis:emojis rendering:EmojiRendering.highQuality];
}

- (void)configureEmojis:(NSDictionary<NSString *, id> *)emojis rendering:(EmojiRendering *)rendering {
    [self applyEmojis:emojis rendering:rendering];
    
    [[NSNotificationCenter defaultCenter] addObserverForName:UITextViewTextDidChangeNotification object:self queue:[NSOperationQueue mainQueue] usingBlock:^(NSNotification * _Nonnull note) {
        [self applyEmojis:emojis rendering:rendering];
    }];
}

- (UIView *)textContainerView {
    return self.subviews[1];
}

- (NSArray<EmojiView *> *)customEmojiViews {
    NSMutableArray<EmojiView *> *views = [NSMutableArray array];
    for (UIView *subview in self.textContainerView.subviews) {
        if ([subview isKindOfClass:[EmojiView class]]) {
            [views addObject:(EmojiView *)subview];
        }
    }
    return views;
}

- (void)applyEmojis:(NSDictionary<NSString *, EmojiSource *> *)emojis rendering:(EmojiRendering *)rendering {
    NSRange range = self.selectedRange;
    NSInteger count = self.attributedText.string.length;
    self.attributedText = [self.attributedText insertEmojis:emojis rendering:rendering];
    NSInteger newCount = self.attributedText.string.length;
    
    for (UIView *view in self.customEmojiViews) {
        [view removeFromSuperview];
    }
    
    [self addEmojiImagesIfNeededWithRendering:rendering];
    
    self.selectedRange = NSMakeRange(range.location - (count - newCount), range.length);
}

- (void)addEmojiImagesIfNeededWithRendering:(EmojiRendering *)rendering {
    [self.attributedText enumerateAttributesInRange:NSMakeRange(0, self.attributedText.length) options:0 usingBlock:^(NSDictionary<NSAttributedStringKey,id> * _Nonnull attributes, NSRange crange, BOOL * _Nonnull stop) {
        dispatch_async(dispatch_get_main_queue(), ^{
            NSTextAttachment *emojiAttachment = attributes[NSAttachmentAttributeName];
            UITextPosition *position1 = [self positionFromPosition:self.beginningOfDocument offset:crange.location];
            UITextPosition *position2 = [self positionFromPosition:position1 offset:crange.length];
            UITextRange *textRange = [self textRangeFromPosition:position1 toPosition:position2];
            NSData *emojiData = emojiAttachment.contents;
            EmojiSource *emoji = nil;
            
            if ([emojiData isKindOfClass:[NSData class]]) {
                @try {
                    emoji = [NSJSONSerialization JSONObjectWithData:emojiData options:0 error:nil];
                } @catch (NSException *exception) {
                    // JSON decoding failed
                }
            }
            
            if (!emoji) {
                return;
            }
            
            CGRect rect = [self firstRectForRange:textRange];
            
            EmojiView *emojiView = [[EmojiView alloc] initWithFrame:rect];
            emojiView.backgroundColor = self.backgroundColor;
            emojiView.userInteractionEnabled = NO;
            
            if ([emoji isKindOfClass:[EmojiSource class]]) {
                switch (emoji.type) {
                    case EmojiSourceTypeCharacter:
                        emojiView.label.text = emoji.emojiValue;
                        break;
                    case EmojiSourceTypeImageUrl:
                        if (![renderViews objectForKey:emoji]) {
                            if (emoji.emojiValue) {
                                NSURL *url = [NSURL URLWithString:emoji.emojiValue];
                                UIImageView *renderView = [[UIImageView alloc] initWithFrame:rect];
                                [renderView setFromURL:url rendering:rendering];
                                [renderViews setObject:renderView forKey:emoji];
                                [self.window addSubview:renderView];
                                renderView.alpha = 0;
                            }
                        }
                        break;
                    case EmojiSourceTypeImageAsset:
                        if (![renderViews objectForKey:emoji]) {
                            UIImageView *renderView = [[UIImageView alloc] initWithFrame:rect];
                            [renderView setFromAsset:emoji.emojiValue rendering:rendering];
                            [renderViews setObject:renderView forKey:emoji];
                            [self.window addSubview:renderView];
                            renderView.alpha = 0;
                        }
                        break;
                    case EmojiSourceTypeAlias:
                        break;
                }
                
                UIImageView *view = [renderViews objectForKey:emoji];
                
                if (view) {
                    [emojiView setFromRenderView:view];
                }
            }
            
            [self.textContainerView addSubview:emojiView];
        });
    }];
}

@end

File: Streamoji/Classes/Extensions/UITextView+Emojis.swift
//
//  UITextView+Emojis.swift
//  Streamoji
//
//  Created by Matheus Cardoso on 30/06/20.
//

import UIKit

fileprivate var renderViews: [EmojiSource: UIImageView] = [:]


// MARK: Public
extension UITextView {
    /// Configures this UITextView to display custom emojis.
    ///
    /// - Parameter emojis: A dictionary of emoji keyed by its shortcode.
    /// - Parameter rendering: The rendering options. Defaults to `.highQuality`.
    public func configureEmojis(_ emojis: [String: EmojiSource], rendering: EmojiRendering = .highQuality) {
        self.applyEmojis(emojis, rendering: rendering)

        NotificationCenter.default.addObserver(
            forName: UITextView.textDidChangeNotification,
            object: self,
            queue: .main
        ) { [weak self] _ in
            self?.applyEmojis(emojis, rendering: rendering)
        }
    }
    
    @objc public func configureEmojis(_ emojis: NSDictionary , quality: Int) {
        var rendering : EmojiRendering = .highQuality;
        switch(quality) {
        case 0:
            rendering = .lowestQuality
            break
        case 1:
            rendering = .lowQuality
            break
        case 2:
            rendering = .mediumQuality
            break
        case 3:
            rendering = .highQuality
            break
        case 4:
            rendering = .highestQuality
            break
        default:
            break;
        }
        
        let dic = emojis as? [String : [String]];
        guard let emojiDic : [String: EmojiSource] = dic?.mapValues({ values in
            if(values.count < 2) {
                return .character("")
            }
            var emoji : EmojiSource
            switch(values[0]) {
            case "character":
                emoji = .character(values[1])
                break
                
            case "imageUrl":
                emoji = .imageUrl(values[1])
                break
                
            case "imageAsset":
                emoji = .imageAsset(values[1])
                break
                
            case "alias":
                emoji = .alias(values[1])
                break
            default:
                emoji = .character(values[1])
                break
            }
            return emoji
        }) else {
            return
        }
        configureEmojis(emojiDic, rendering: rendering)
    }
}

// MARK: Private
extension UITextView {
    private var textContainerView: UIView { subviews[1] }
    
    private var customEmojiViews: [EmojiView] {
        textContainerView.subviews.compactMap { $0 as? EmojiView }
    }
    
    private func applyEmojis(_ emojis: [String: EmojiSource], rendering: EmojiRendering) {
        let range = selectedRange
        let count = attributedText?.string.count ?? 0
        self.attributedText = attributedText.insertingEmojis(emojis, rendering: rendering)
        let newCount = attributedText.string.count
        customEmojiViews.forEach { $0.removeFromSuperview() }
        addEmojiImagesIfNeeded(rendering: rendering)
        selectedRange = NSRange(location: range.location - (count - newCount), length: range.length)
    }
    
    private func addEmojiImagesIfNeeded(rendering: EmojiRendering) {
        attributedText.enumerateAttributes(in: NSRange(location: 0, length: attributedText.length), options: [], using: { attributes, crange, _ in
            DispatchQueue.main.async {
                guard
                    let emojiAttachment = attributes[NSAttributedString.Key.attachment] as? NSTextAttachment,
                    let position1 = self.position(from: self.beginningOfDocument, offset: crange.location),
                    let position2 = self.position(from: position1, offset: crange.length),
                    let range = self.textRange(from: position1, to: position2),
                    let emojiData = emojiAttachment.contents,
                    let emoji = try? JSONDecoder().decode(EmojiSource.self, from: emojiData)
                else {
                    return
                }
                
                let rect = self.firstRect(for: range)

                let emojiView = EmojiView(frame: rect)
                emojiView.backgroundColor = self.backgroundColor
                emojiView.isUserInteractionEnabled = false
                
                switch emoji {
                case let .character(character):
                    emojiView.label.text = character
                case let .imageUrl(imageUrl):
                    guard renderViews[emoji] == nil else {
                        break
                    }
                    
                    if let url = URL(string: imageUrl) {
                        let renderView = UIImageView(frame: rect)
                        renderView.setFromURL(url, rendering: rendering)
                        renderViews[emoji] = renderView
                        self.window?.addSubview(renderView)
                        renderView.alpha = 0
                    }
                case let .imageAsset(imageAsset):
                    guard renderViews[emoji] == nil else {
                        break
                    }
                    
                    let renderView = UIImageView(frame: rect)
                    renderView.setFromAsset(imageAsset, rendering: rendering)
                    renderViews[emoji] = renderView
                    self.window?.addSubview(renderView)
                    renderView.alpha = 0
                case .alias:
                    break
                }
                
                if let view = renderViews[emoji] {
                    emojiView.setFromRenderView(view)
                }
                
                self.textContainerView.addSubview(emojiView)
            }
        })
    }
}
